static-analysis/lint/src/main/kotlin/com/uber/autodispose/lint/AutoDisposeDetector.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2018. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.autodispose.lint
+
+import com.android.tools.lint.client.api.JavaEvaluator
+import com.android.tools.lint.detector.api.*
+import com.intellij.psi.PsiMethod
+import org.jetbrains.uast.UCallExpression
+import java.util.EnumSet
+
+/**
+ * Detector which checks if your stream subscriptions are handled by AutoDispose.
+ */
+class AutoDisposeDetector: Detector(), SourceCodeScanner {
+
+  companion object {
+    val ISSUE: Issue = Issue.create(
+        "AutoDisposeUsage",
+        "Subscription not managed with AutoDispose",
+        "You're subscribing to an observable but not handling it's subscription. This " +
+            "can result in memory leaks. You can avoid memory leaks by appending " +
+            "`.as(autoDisposable(this))` before you subscribe.",
+        Category.CORRECTNESS,
+        8,
+        Severity.ERROR,
+        Implementation(AutoDisposeDetector::class.java, EnumSet.of(Scope.JAVA_FILE, Scope.TEST_SOURCES)))
+
+    private const val OBSERVABLE = "io.reactivex.Observable"
+    private const val FLOWABLE = "io.reactivex.Flowable"
+    private const val SINGLE = "io.reactivex.Single"
+    private const val MAYBE = "io.reactivex.Maybe"
+    private const val COMPLETABLE = "io.reactivex.Completable"
+  }
+
+  override fun getApplicableMethodNames(): List<String> = listOf("subscribe")
+
+  override fun visitMethod(context: JavaContext, node: UCallExpression, method: PsiMethod) {
+    val evaluator = context.evaluator
+
+    if (isReactiveType(evaluator, method)) {
+      context.report(ISSUE, node, context.getLocation(node), "Subscription not managed by AutoDispose.")
+    }
+  }
+
+  private fun isReactiveType(evaluator: JavaEvaluator, method: PsiMethod): Boolean {
+    return evaluator.isMemberInClass(method, OBSERVABLE) ||
```

Comment:
There's a bit more to be done here. Specifically, we want to _only_ apply this in scopes that have lifecycles. The error prone check does this as well if you want some prior art. Basically:

- `isReactiveType`
- "is subscribing inside a class that implements "known" lifecycle types
  - in the EP checker, we include a few "known" lifecycle types like activity/fragment/scopeprovider/etc

Then, potential icing on the cake, allow contributions of custom lifecycle types. I'm not really sure how feasible that is in lint though, so not something we need to do in this PR

Reply:
Should it only be applied for scopes that have lifecycles? For example, if you're using a Presenter/ViewModel/Controller etc. those are all still valid cases where lifecycle should be handled.

Reply:
Yep, only places that have scopes. Those cases would either need to implement ScopeProvider or (if we can accept custom types) specified separately